File: scripts/python/csv_parse.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
import matplotlib.pyplot as plt

csvfilename = '../sample_sensor_data.csv'
axis_names = ['X', 'Y', 'Z']

# Rows:
# 0 - Timestamp
# 1 - Accel X
# 2 - Accel Y
# 3 - Accel Z
# 4 - Gyro X
# 5 - Gyro Y
# 6 - Gyro Z
# 7 - NaNs

data = np.genfromtxt(csvfilename, delimiter=',').T

timestamps = (data[0] - data[0, 0]) / 1000  # Normalized

accel_data = data[1:4]
gyro_data = data[4:-1]  # Deal with that trailing comma

# Plot Accelerometer Data
fig, axarr = plt.subplots(3)
fig.tight_layout()
fig.subplots_adjust(top=0.88)
fig.suptitle('Accelerometer Data', size=16)
for i in range(3):
    axarr[i].plot(timestamps, accel_data[i])
    axarr[i].set_title('Accelerometer : {}'.format(axis_names[i]))

plt.savefig('accel_data.png')

# Plot Gyroscope Data
fig, axarr = plt.subplots(3)
fig.tight_layout()
fig.subplots_adjust(top=0.88)
fig.suptitle('Gyroscope Data', size=16)
for i in range(3):
    axarr[i].plot(timestamps, gyro_data[i])
    axarr[i].set_title('Gyroscope : {}'.format(axis_names[i]))

plt.savefig('gyro_data.png')

# Generate filter


def filter_gen(Fs=25, Fc=12, transBand=4, numtaps=10):
    """ Generate a filter, with the given parameters. """
    import scipy.signal as ssig

    # Check out some more right here:
    # http://docs.scipy.org/doc/scipy/reference/signal.html#filter-design
    h = ssig.firwin(numtaps=numtaps, cutoff=Fc / Fs, width=transBand / Fs)
    return h


def view_filter(b):
    """ Given a filter, look at it. Like MATLAB's freqz.
        http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.signal.freqz.html
    """
    import scipy.signal as ssig
    w, h = ssig.freqz(b)

    fig = plt.figure()
    plt.title('Digital filter frequency response')
    ax1 = fig.add_subplot(111)

    plt.plot(w, 20 * np.log10(abs(h)), 'b')
    plt.ylabel('Amplitude [dB]', color='b')
    plt.xlabel('Frequency [rad/sample]')

    ax2 = ax1.twinx()
    angles = np.unwrap(np.angle(h))
    plt.plot(w, angles, 'g')
    plt.ylabel('Angle (radians)', color='g')
    plt.grid()
    plt.axis('tight')
    plt.savefig('designed_filter.png')


Fs = 100        # Sampling Rate
Fc = 12         # Cutoff Frequency
transBand = 10  # Width of transition band
numtaps = 10    # Number of taps

b = filter_gen(Fs, Fc, transBand, numtaps)
view_filter(b)

File: scripts/matlab/template/csv_parse.m
%% CSV Parser
% Dario Aranguiz

csvFile = 'C:\Users\daran\Dropbox\Petronics\sensor_data_dario_house.csv';
axis_names = ['X', 'Y', 'Z'];

% Rows: 
% 1 - Timestamp
% 2 - Accel X
% 3 - Accel Y
% 4 - Accel Z
% 5 - Gyro X
% 6 - Gyro Y
% 7 - Gyro Z
% 8 - Orientation X
% 9 - Orientation Y
% 10 - Orientation Z
csv = csvread(csvFile);

accel = csv(:, 2:4);
gyro = csv(:, 5:7);
orien = mod(180 + csv(:, end-3:end-1), 360);

% Normalize timestamps
timestamps = csv(:, 1) - csv(1,1);
timestamps = timestamps / 1000;

% Plot accelerometer
figure;
for i=1:3
    subplot(3,1,i), plot(timestamps, accel(:, i));
    title(strcat('Accel ', axis_names(i)));
end
xlabel('Time in seconds');

% Plot gyroscope
figure;
title('Gyroscope Readings');
for i=1:3
    subplot(3,1,i), plot(timestamps, gyro(:, i));
    title(strcat('Gyro ', axis_names(i)));
end
xlabel('Time in seconds');

figure;
title('Orientation Readings');
for i=1:3
    subplot(3,1,i), plot(timestamps, orien(:, i));
    title(strcat('Orientation ', axis_names(i)));
end
xlabel('Time in seconds');

%% Generate filter

% TODO: Select appropriate filter parameters 

Fs = 1000;           % Sampling Rate
Fc = 500;           % Cutoff Frequency
transBand = 50;       % Width of transition band
n = 100;             % Number of taps 

b = filter_gen(Fs, Fc, transBand, n);

%% Resample

% http://www.mathworks.com/help/signal/ref/resample.html
[accelResampled, accelTime] = resample(accel, timestamps, Fs);

figure;
for i=1:3
    subplot(3,1,i), plot(accelTime, accelResampled(:, i));
    title(strcat('Accel Resampled ', axis_names(i)));
end
xlabel('Time in seconds');

% TODO: Resample gyroscope

%% Apply filter

accelFiltered = filter(b, 1, accelResampled);

figure;
for i=1:3
    subplot(3,1,i), plot(accelTime, accelFiltered(:, i));
    title(strcat('Accel Filtered ', axis_names(i)));
end
xlabel('Time in seconds');

% TODO: Filter accelerometer
